File: Recursion.py
class Tree_Recursion_Class:
    def __init__(self):
        self.call_count = 0
        self.current_depth = 0
        self.max_depth = 0

    def initialize_call_count(self):
        # Initializer Method
        self.call_count = 0
        self.current_depth = 0
        self.max_depth = 0

    def tree_recursion(self, n):
        # Tree Recursion Implementation Method
        print(n, end=", ")

        self.call_count += 1
        self.current_depth += 1
        self.max_depth = max(self.current_depth, self.max_depth)
        if n > 0:
            self.tree_recursion(n - 1)
            self.tree_recursion(n - 1)
        self.current_depth -= 1

    def get_call_count(self):
        # Getter Method
        return self.call_count

    def get_max_depth(self):
        # Getter Method
        return self.max_depth

class Natural_Numbers_Class:
    def __init__(self):
        self.recursive_stack_count = 0
        self.analytic_stack_count = 0

    def recursive_sum(self, n):
        self.recursive_stack_count += 1
        if (n == 1):
            return 1
        return n + self.recursive_sum(n-1)
    def analytic_sum(self, n):
        return (n*(n + 1)/2)

class Power_Series_Class:
    def __init__(self):
        pass

    def recursive_method_one(self, m, n):
        if (n == 0):
            return 1
        return (m)*self.recursive_method_one(m,n-1)

    def recursive_method_two(self, m, n):
        if (n ==0):
            return 1
        if (n % 2 == 0):
            return self.recursive_method_two(m*m,n/2)
        else:
            return m*self.recursive_method_two(m*m, (n-1)/2)

class Taylor_Series_Class:
    def __init__(self):
        pass

    p = 1
    f = 1
    @staticmethod
    def taylor_method_one(x, n):
        if n == 0:
            return 1
        r = Taylor_Series_Class.taylor_method_one(x, n - 1)
        Taylor_Series_Class.p = Taylor_Series_Class.p*x
        Taylor_Series_Class.f = Taylor_Series_Class.f*n
        return r + (Taylor_Series_Class.p / Taylor_Series_Class.f)

    s = 1
    @staticmethod
    def taylor_series_horner_method(x, n):
        if n == 0:
            return Taylor_Series_Class.s
        Taylor_Series_Class.s = 1 + (x / n) * Taylor_Series_Class.s
        return Taylor_Series_Class.taylor_series_horner_method(x, n - 1)

class Fibonacci_Series_Class:
    def __init__(self):
        pass

    def fibonacci_method_one(self, n):
        if (n <=1):
            return n
        else:
            return (self.fibonacci_method_one(n-2) + self.fibonacci_method_one(n-1))

    F = []

    def fibonacci_method_two_list(self,n):
        Fibonacci_Series_Class.F = [-1] * n

    def fibonacci_method_two(self, n):
        if n <= 1:
            return n
        if Fibonacci_Series_Class.F[n-1] != -1:
            return Fibonacci_Series_Class.F[n-1]
        Fibonacci_Series_Class.F[n-1] = self.fibonacci_method_two(n - 1) + self.fibonacci_method_two(n - 2)
        return Fibonacci_Series_Class.F[n-1]

class Binomial_Method():
    def __init__(self):
        pass
    def factorial(self, n):
        if (n <=1):
            return n
        else:
            return n*self.factorial(n-1)
    def recursive_method(self, n, r):
        num = self.factorial(n)
        denom_a = self.factorial(r)
        denom_b = self.factorial(n-r)
        return (num/(denom_a*denom_b))
    def pascal_method(self, n,r):
        if (n == r or r == 0):
            return 1
        else:
            return (self.pascal_method(n - 1, r - 1) + self.pascal_method(n - 1, r))

class Tower_of_Hanoi():
    def __init__(self):
        pass
    def TOH(self, a,b,c,n):
        if (n>0):
            self.TOH( a, c, b, n-1)
            print("Move disk from tower ", a, " to tower ", c)
            self.TOH( b, a, c, n-1)

File: Recursion.cpp
#include "Recursion.h"
void Recursion::initialize_call_count()
{
	//Initiliser Method 
	call_count = 0;
	current_depth = 0;
	max_depth = 0;
}
void Recursion::tree_recursion(int n)
{
	// Tree Recursion Implementation Method
	std::cout << n << ", ";

	call_count++;
	current_depth++;
	max_depth = std::max(current_depth, max_depth);
	if (n > 0)
	{
		tree_recursion(n - 1);
		tree_recursion(n - 1);
	}
	current_depth--;
}
//Getter Methods
int Recursion::get_call_count()
{
	return call_count;
}
int Recursion::get_max_depth()
{
	return max_depth;
}
//Natural Numbers
int Recursion::recursive_sum(int N)
{
	recursive_stack_count++;
	if (N == 1)
	{
		return 1;
	}
	else
	{
		return (recursive_sum(N - 1) + N);
	}
}
int Recursion::analytical_sum(int N)
{
	analytical_stack_count++;
	return ((N * (N + 1)) / 2);
}
void Recursion::initialize_counts()
{
	recursive_stack_count = 0;
	analytical_stack_count = 0;
}
//Power Series
void Recursion::initilize_power_counts()
{
	power_stack_count = 0;
	improved_power_stack_count = 0;
}
int Recursion::power_recursion(int m, int n)
{
	power_stack_count++;
	if (n == 0)
	{
		return 1;
	}
	else
	{
		return (m*power_recursion(m,n-1));
	}
}
int Recursion::improved_power_recursion(int m, int n)
{
	improved_power_stack_count++;
	if (n == 0)
	{
		return 1;
	}
	else if (n % 2 == 0)
	{
		return (improved_power_recursion(m*m, n/2));
	}
	else
	{
		return (m*improved_power_recursion(m*m,(n-1)/2));
	}
}
//Taylor Series with Static Variables
double Recursion::taylor_method_one(int x, int n)
{
	static double p=1, f = 1;
	if (n == 0)
	{
		return 1;
	}
	else
	{
		double r = taylor_method_one(x, (n - 1));
		p = p * x;
		f = f * n;
		return (r + (p / f));
	}
}
double Recursion::taylor_series_horner_method(int x, int n)
{
	static double s=1;
	if (n == 0)
	{
		return s;
	}
	s = 1 + (x / static_cast<double>(n))*s;
	return taylor_series_horner_method(x, n - 1);
}
// Fibonacci Series
int Recursion::fibonacci_method_one(int n)
{
	fib_m1_stack++;
	if (n <= 1)
	{
		return n;
	}
	else
	{
		return (fibonacci_method_one(n - 2) + fibonacci_method_one(n - 1));
	}
}
void Recursion::fibonacci_method_two_vector(int n)
{
	fib_m1_stack = 0;
	fib_m2_stack = 0;
	F = std::vector<int>(n, -1);
}
int Recursion::fibonacci_method_two(int n)
{
	fib_m2_stack++;
	if (n <= 1) {
        return n;
    }
	if (F[n] != -1) {
		return F[n];
	}
	F[n] = fibonacci_method_two(n - 1) + fibonacci_method_two(n - 2);
	return F[n];
}
// nCr Analysis
void Recursion::initialize_stacks_nCr()
{
	pascal_stack = 0;
	factorial_stack = 0;
}
int Recursion::factorial(int n)
{
	factorial_stack++;
	if (n == 1)
	{
		return 1;
	}
	else
	{
		return (n * factorial(n - 1));
	}
}
int Recursion::factorial_method(int n, int r)
{
	int num = factorial(n);
	int denom_a = factorial(r);
	int denom_b = factorial(n - r);
	return (num / (denom_a * denom_b));
}
int Recursion::pascal_triangle_method(int n, int r)
{
	pascal_stack++;
	if (n == r || r == 0)
	{
		return 1;
	}
	return (pascal_triangle_method(n-1,r-1) + pascal_triangle_method(n-1,r));
}
// TOH
void Recursion::TOH(int n, int A, int B, int C)
{
	if (n > 0)
	{
		TOH(n - 1, A, C, B);
		std::cout << "\nMove disk from tower " << A << " to tower " << C;
		TOH(n - 1, B, A, C);
	}
}